src/Microsoft.AspNetCore.Sockets.Common/Formatters/ServerSentEventsMessageFormatter.cs

```diff
@@ -0,0 +1,186 @@
+                throw new InvalidOperationException("Cannot format message where endOfMessage is false using this format");
+            }
+
+            // Need at least: Length of 'data: ', one character type, one \r\n, and the trailing \r\n
+            if (buffer.Length < DataPrefix.Length + 1 + Newline.Length + Newline.Length)
+            {
+                bytesWritten = 0;
+                return false;
+            }
+            DataPrefix.CopyTo(buffer);
+            buffer = buffer.Slice(DataPrefix.Length);
+            if (!TryFormatType(buffer, message.Type))
+            {
+                bytesWritten = 0;
+                return false;
+            }
+            buffer = buffer.Slice(1);
+
+            Newline.CopyTo(buffer);
+            buffer = buffer.Slice(Newline.Length);
+
+            // Write the payload
+            if (!TryFormatPayload(message.Payload.Buffer, message.Type, buffer, out var writtenForPayload))
+            {
+                bytesWritten = 0;
+                return false;
+            }
+            buffer = buffer.Slice(writtenForPayload);
+
+            if (buffer.Length < Newline.Length)
+            {
+                bytesWritten = 0;
+                return false;
+            }
+            Newline.CopyTo(buffer);
+
+            bytesWritten = DataPrefix.Length + Newline.Length + 1 + writtenForPayload + Newline.Length;
+            return true;
+        }
+
+        private static bool TryFormatPayload(ReadableBuffer payload, MessageType type, Span<byte> buffer, out int bytesWritten)
+        {
+            // Short-cut for empty payload
+            if (payload.Length == 0)
+            {
+                bytesWritten = 0;
+                return true;
+            }
+
+            var writtenSoFar = 0;
+            if (type == MessageType.Binary)
+            {
+                // TODO: We're going to need to fix this as part of https://github.com/aspnet/SignalR/issues/192
+                var message = Convert.ToBase64String(payload.ToArray());
+                var encodedSize = DataPrefix.Length + Encoding.UTF8.GetByteCount(message) + Newline.Length;
+                if (buffer.Length < encodedSize)
+                {
+                    bytesWritten = 0;
+                    return false;
+                }
+                DataPrefix.CopyTo(buffer);
+                buffer = buffer.Slice(DataPrefix.Length);
+
+                var array = Encoding.UTF8.GetBytes(message);
+                array.CopyTo(buffer);
+                buffer = buffer.Slice(array.Length);
+
+                Newline.CopyTo(buffer);
+                writtenSoFar += encodedSize;
+                buffer.Slice(Newline.Length);
+            }
+            else
+            {
+                while (true)
+                {
+                    // Seek to the end of buffer or newline
+                    var sliced = payload.TrySliceTo(LineFeed, out var slice, out var cursor);
```

Comment:
Yeah, I'm going to do that separately since I had already written this code and the tests when we decided that.